### PR TITLE
add types for middleware

### PR DESCRIPTION
**Pull request by mloberg**
While working with middleware functions, it would be nice to reference a type rather than having to copy/paste the whole function.

Thanks for the work on Resty!

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -66,6 +66,15 @@ var (
 	bufPool           = &sync.Pool{New: func() interface{} { return &bytes.Buffer{} }}
 )
 
+type (
```
Comment (jeevatkm) [inline client.go:69]: @mloberg Can you please add godoc for the exported types?

Comment (mloberg) [reply client.go:69]: added

File: client.go [hunk 2]
```diff
@@ -66,6 +66,23 @@ var (
...
+	// RequestMiddleware type is for request middleware, called before a request is sent
+	RequestMiddleware func(*Client, *Request) error
+
+	// ResponseMiddleware type is for response middleware, called after a response has been recieved
```
Comment (moorereason) [inline client.go:73]: s/recieved/received/

Comment (mloberg) [reply client.go]: :derp: fixed

**Pull request merged**